### ISSUE TITLE
[Query] use share memory in a simple query: Query23

### ISSUE THREAD

**Issue by JigaoLuo** [cakebytheoceanLuo/dogqc#7]
0. try best to use only one kernel
  - almost impossible using only one kernel: https://stackoverflow.com/questions/31718712/use-cudadevicesynchronize-inside-kernel-for-global-synchronisation  https://forums.developer.nvidia.com/t/device-synchronization-inside-cuda-kernels/45261
1. use multiple hash tables instead of using a global one
2. each SM finds its own hash table (maybe on each SM's inclusive L1)
3. try to do some share memory

TODO:
- set number of grids. Now it seems 1.

Lesson Learn:
- It is enough to let hash tables have size 128 for each block.
  - hash aggregation is a reduce operation, which is not the case of hash join on keys.

**Comment by JigaoLuo** [comment]
# SQL

```sql
select l_linenumber, count(*)
from lineitem 
group by l_linenumber
```

![image](https://user-images.githubusercontent.com/28505034/145018713-c34816ff-1973-4fa3-944e-5d917e27d0c2.png)


## DogQC

### Pseudocode of CUDA

```
cudaMalloc&init hash table, hash group (an array), output stuff;
cudaDeviceSynchronize();
cudaMemcpy lineitem table's inputs;
cudaDeviceSynchronize();

/// Kernel1: build hash table and hash groups. each thread takes a tuple of lineitem
krnl_lineitem1<<<920,128>>>(input, hash table, hash group);

cudaDeviceSynchronize();

/// Kernel2: lookup hash table and hash groups. each thread takes an entry in hash table and hash group
krnl_aggregation2<<<920,128>>>(hash table, hash group, output projections);
cudaDeviceSynchronize();
```

### Execution

```
$ ./run_single_query.sh 23
compilation...
nvcc tpch23.cu -o tpch23 -std=c++11 -arch=sm_75  -Xcompiler="-pthread " 
compilation time: 1416.3 ms

execution...
Memory 0.2 / 5.8 GB
Memory 0.5 / 5.8 GB

Result: 7 tuples
l_linenumber:        2  count_l_linenumber:  1285828  
l_linenumber:        3  count_l_linenumber:  1071394  
l_linenumber:        7  count_l_linenumber:   214621  
l_linenumber:        5  count_l_linenumber:   643287  
l_linenumber:        6  count_l_linenumber:   429070  
l_linenumber:        4  count_l_linenumber:   857015  
l_linenumber:        1  count_l_linenumber:  1500000  

<timing>
                  krnl_lineitem1:    3.5 ms
               krnl_aggregation2:    1.7 ms
                          finish:    0.0 ms
                 totalKernelTime:    5.2 ms
</timing>

```

### Profiling

#### TL;DR
```
24.28%  3.6181ms         1  3.6181ms  3.6181ms  3.6181ms  krnl_lineitem1(int*, agg_ht<apayl2>*, int*)
11.13%  1.6577ms         1  1.6577ms  1.6577ms  1.6577ms  krnl_aggregation2(agg_ht<apayl2>*, int*, int*, int*, int*)
```

#### Full Output

```
$ nvprof ./tpch23
==4227== NVPROF is profiling process 4227, command: ./tpch23
Memory 0.2 / 5.8 GB
Memory 0.5 / 5.8 GB

Result: 7 tuples
l_linenumber:        2  count_l_linenumber:  1285828  
l_linenumber:        3  count_l_linenumber:  1071394  
l_linenumber:        7  count_l_linenumber:   214621  
l_linenumber:        5  count_l_linenumber:   643287  
l_linenumber:        6  count_l_linenumber:   429070  
l_linenumber:        4  count_l_linenumber:   857015  
l_linenumber:        1  count_l_linenumber:  1500000  

<timing>
                  krnl_lineitem1:    3.6 ms
               krnl_aggregation2:    1.7 ms
                          finish:    0.0 ms
                 totalKernelTime:    5.3 ms
</timing>
==4227== Profiling application: ./tpch23
==4227== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   30.07%  4.4804ms         3  1.4935ms  1.4720us  2.2469ms  [CUDA memcpy DtoH]
                   24.28%  3.6181ms         1  3.6181ms  3.6181ms  3.6181ms  krnl_lineitem1(int*, agg_ht<apayl2>*, int*)
                   16.96%  2.5275ms         1  2.5275ms  2.5275ms  2.5275ms  [CUDA memcpy HtoD]
                   15.50%  2.3103ms         1  2.3103ms  2.3103ms  2.3103ms  void initAggHT<apayl2>(agg_ht<apayl2>*, int)
                   11.13%  1.6577ms         1  1.6577ms  1.6577ms  1.6577ms  krnl_aggregation2(agg_ht<apayl2>*, int*, int*, int*, int*)
                    2.06%  306.68us         2  153.34us  2.5930us  304.08us  void initArray<int>(int*, int, int)
      API calls:   95.83%  207.02ms         8  25.878ms  1.4620us  199.14ms  cudaDeviceSynchronize
                    3.39%  7.3131ms         4  1.8283ms  12.683us  2.6454ms  cudaMemcpy
                    0.31%  677.10us         6  112.85us  60.121us  290.70us  cudaMalloc
                    0.23%  498.65us         6  83.109us  37.429us  216.68us  cudaFree
                    0.09%  191.93us         1  191.93us  191.93us  191.93us  cuDeviceTotalMem
                    0.08%  181.27us        97  1.8680us     161ns  75.886us  cuDeviceGetAttribute
                    0.03%  67.052us         5  13.410us  3.2690us  24.413us  cudaLaunchKernel
                    0.01%  31.843us         2  15.921us  14.112us  17.731us  cudaMemGetInfo
                    0.01%  29.142us         1  29.142us  29.142us  29.142us  cuDeviceGetName
                    0.00%  9.0300us         1  9.0300us  9.0300us  9.0300us  cuDeviceGetPCIBusId
                    0.00%  1.5520us         8     194ns     125ns     404ns  cudaGetLastError
                    0.00%  1.4440us         3     481ns     176ns  1.0020us  cuDeviceGetCount
                    0.00%     947ns         2     473ns     156ns     791ns  cuDeviceGet
                    0.00%     250ns         1     250ns     250ns     250ns  cuDeviceGetUuid
```
![image](https://user-images.githubusercontent.com/28505034/144736318-81540248-76ba-4e38-b473-035199f4c777.png)

**Comment by JigaoLuo** [comment]
## Shared Memory DogQC (Handcoded)

### Pseudocode

```
cudaMalloc&init output stuff;
cudaDeviceSynchronize();
cudaMemcpy lineitem table's inputs;
cudaDeviceSynchronize();

/// Kernel
krnl_lineitem1<<<920,128>>>(input, output projections);
cudaDeviceSynchronize();

output is block-independent 
reduce (aggregate) output on CPU /// now is single threaded
```

The detail inside of the kernel code:

```
const int HT_SIZE = 128;
__shared__ agg_ht<apayl2> aht2[HT_SIZE];
__shared__ int agg1[HT_SIZE];
init hash table and hash group
__syncthreads();

/// similar as kernel 1
build hash table&group upon lineitem
__syncthreads();

/// similar as kernel 2
lookup hash table&group
```

### Execution

```
$ ./tpch23_advanced
Memory 0.2 / 5.8 GB
Memory 0.2 / 5.8 GB

Result: 6440 tuples
l_linenumber:        5  count_l_linenumber:      703  
l_linenumber:        3  count_l_linenumber:     1161  
l_linenumber:        7  count_l_linenumber:      230  
l_linenumber:        1  count_l_linenumber:     1630  
l_linenumber:        4  count_l_linenumber:      928  
l_linenumber:        6  count_l_linenumber:      483  
l_linenumber:        2  count_l_linenumber:     1386  
l_linenumber:        5  count_l_linenumber:      701  
l_linenumber:        3  count_l_linenumber:     1160  
l_linenumber:        7  count_l_linenumber:      253  
[...]

MY REDUCE ON CPU (single-cpu-threaded)
l_linenumber:        2  count_l_linenumber:  1285828  
l_linenumber:        3  count_l_linenumber:  1071394  
l_linenumber:        5  count_l_linenumber:   643287  
l_linenumber:        7  count_l_linenumber:   214621  
l_linenumber:        1  count_l_linenumber:  1500000  
l_linenumber:        4  count_l_linenumber:   857015  
l_linenumber:        6  count_l_linenumber:   429070  
<timing>
                          finish:    0.0 ms
                 totalKernelTime:    0.6 ms
reduce on CPU (single-cpu-threaded):    0.5 ms
</timing>
```

` 6440 tuples` = `7 tuples` * `920 blocks`

### Profiling

#### TL;DR
```
8.16%  614.52us         1  614.52us  614.52us  614.52us  krnl_lineitem1(int*, int*, int*, int*)
```

#### Full Output

```
$ nvprof ./tpch23_advanced
==16046== NVPROF is profiling process 16046, command: ./tpch23_advanced
Memory 0.2 / 5.8 GB
Memory 0.2 / 5.8 GB

Result: 6440 tuples
l_linenumber:        6  count_l_linenumber:      476  
l_linenumber:        2  count_l_linenumber:     1383  
l_linenumber:        5  count_l_linenumber:      705  
l_linenumber:        3  count_l_linenumber:     1147  
l_linenumber:        7  count_l_linenumber:      233  
l_linenumber:        1  count_l_linenumber:     1652  
l_linenumber:        4  count_l_linenumber:      932  
l_linenumber:        5  count_l_linenumber:      701  
l_linenumber:        6  count_l_linenumber:      457  
l_linenumber:        2  count_l_linenumber:     1419  
[...]

MY REDUCE ON CPU (single-cpu-threaded)
l_linenumber:        4  count_l_linenumber:   857015  
l_linenumber:        2  count_l_linenumber:  1285828  
l_linenumber:        6  count_l_linenumber:   429070  
l_linenumber:        5  count_l_linenumber:   643287  
l_linenumber:        3  count_l_linenumber:  1071394  
l_linenumber:        7  count_l_linenumber:   214621  
l_linenumber:        1  count_l_linenumber:  1500000  
<timing>
                          finish:    0.1 ms
                 totalKernelTime:    0.6 ms
reduce on CPU (single-cpu-threaded):    0.6 ms
</timing>
==16046== Profiling application: ./tpch23_advanced
==16046== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   57.89%  4.3596ms         3  1.4532ms  1.1200us  2.1874ms  [CUDA memcpy DtoH]
                   33.92%  2.5543ms         1  2.5543ms  2.5543ms  2.5543ms  [CUDA memcpy HtoD]
                    8.16%  614.52us         1  614.52us  614.52us  614.52us  krnl_lineitem1(int*, int*, int*, int*)
                    0.03%  2.3690us         1  2.3690us  2.3690us  2.3690us  void initArray<int>(int*, int, int)
      API calls:   96.00%  198.70ms         7  28.385ms  1.7060us  198.00ms  cudaDeviceSynchronize
                    3.49%  7.2212ms         4  1.8053ms  12.021us  2.6759ms  cudaMemcpy
                    0.15%  307.00us         4  76.750us  56.553us  109.23us  cudaMalloc
                    0.10%  216.53us         4  54.131us  39.173us  72.187us  cudaFree
                    0.09%  191.76us         1  191.76us  191.76us  191.76us  cuDeviceTotalMem
                    0.09%  180.90us        97  1.8640us     165ns  75.085us  cuDeviceGetAttribute
                    0.04%  74.220us         2  37.110us  24.684us  49.536us  cudaLaunchKernel
                    0.02%  43.844us         2  21.922us  18.056us  25.788us  cudaMemGetInfo
                    0.01%  28.444us         1  28.444us  28.444us  28.444us  cuDeviceGetName
                    0.00%  8.8920us         1  8.8920us  8.8920us  8.8920us  cuDeviceGetPCIBusId
                    0.00%  1.5110us         3     503ns     183ns     983ns  cuDeviceGetCount
                    0.00%  1.1850us         7     169ns      98ns     416ns  cudaGetLastError
                    0.00%     917ns         2     458ns     197ns     720ns  cuDeviceGet
                    0.00%     287ns         1     287ns     287ns     287ns  cuDeviceGetUuid
```

![image](https://user-images.githubusercontent.com/28505034/144736309-6112c004-8e6b-4f39-913f-21997cc88777.png)

**Comment by JigaoLuo** [comment]
# Shared Memory DogQC (Handcoded) + Reduce on GPU

**Assumption: The Query Optimizer knows the size of hash table.**

## Pseudocode
```
cudaMalloc&init output stuff;
cudaDeviceSynchronize();
cudaMemcpy lineitem table's inputs;
cudaDeviceSynchronize();

/// Kernel
krnl_lineitem1<<<920,128>>>(input, output projections);
cudaDeviceSynchronize();

kernel_reduce1<<<920,128>>>(...);  /// Build hash table & groups
cudaDeviceSynchronize();
kernel_reduce2<<<920,128>>>(...);  /// Lookup hash table & groups
cudaDeviceSynchronize();
```

## Execution
```
(base) jigao@jigao-System-Product-Name:~/Desktop/dogqc/bin$ ./tpch23_advanced 
Memory 0.2 / 5.8 GB
Memory 0.2 / 5.8 GB
Shared memory usage: 3584 / 49152 bytes.

Result: 6440 tuples
l_linenumber:        3  count_l_linenumber:  1071394  
l_linenumber:        6  count_l_linenumber:   429070  
l_linenumber:        5  count_l_linenumber:   643287  
l_linenumber:        1  count_l_linenumber:  1500000  
l_linenumber:        4  count_l_linenumber:   857015  
l_linenumber:        2  count_l_linenumber:  1285828  
l_linenumber:        7  count_l_linenumber:   214621  
l_linenumber:        0  count_l_linenumber:        0  
l_linenumber:        0  count_l_linenumber:        0  
l_linenumber:        0  count_l_linenumber:        0  
[...]


Final Reduced Result: 7 tuples
l_linenumber:        3  count_l_linenumber:  1071394  
l_linenumber:        6  count_l_linenumber:   429070  
l_linenumber:        5  count_l_linenumber:   643287  
l_linenumber:        1  count_l_linenumber:  1500000  
l_linenumber:        4  count_l_linenumber:   857015  
l_linenumber:        2  count_l_linenumber:  1285828  
l_linenumber:        7  count_l_linenumber:   214621  

<timing>
                          finish:    0.1 ms
                 totalKernelTime:    0.7 ms
                    krnl_reduce1:    0.0 ms
                    krnl_reduce2:    0.0 ms
</timing>
```

## Profiling

### TL;DR

```
9.05%  636.74us         1  636.74us  636.74us  636.74us  krnl_lineitem1(int*, int*, int*, int*)
0.09%  6.3680us         1  6.3680us  6.3680us  6.3680us  krnl_reduce1(int*, int*, agg_ht<apayl2>*, int*)
0.09%  6.3680us         1  6.3680us  6.3680us  6.3680us  krnl_reduce2(agg_ht<apayl2>*, int*, int*, int*, int*)
```

### Full Output

```
(base) jigao@jigao-System-Product-Name:~/Desktop/dogqc/bin$ nvprof ./tpch23_advanced
==13514== NVPROF is profiling process 13514, command: ./tpch23_advanced
Memory 0.2 / 5.8 GB
Memory 0.2 / 5.8 GB
Shared memory usage: 3584 / 49152 bytes.

Result: 6440 tuples
l_linenumber:        7  count_l_linenumber:   214621  
l_linenumber:        3  count_l_linenumber:  1071394  
l_linenumber:        1  count_l_linenumber:  1500000  
l_linenumber:        4  count_l_linenumber:   857015  
l_linenumber:        5  count_l_linenumber:   643287  
l_linenumber:        6  count_l_linenumber:   429070  
l_linenumber:        2  count_l_linenumber:  1285828  
l_linenumber:        0  count_l_linenumber:        0  
l_linenumber:        0  count_l_linenumber:        0  
l_linenumber:        0  count_l_linenumber:        0  
[...]


Final Reduced Result: 7 tuples
l_linenumber:        7  count_l_linenumber:   214621  
l_linenumber:        3  count_l_linenumber:  1071394  
l_linenumber:        1  count_l_linenumber:  1500000  
l_linenumber:        4  count_l_linenumber:   857015  
l_linenumber:        5  count_l_linenumber:   643287  
l_linenumber:        6  count_l_linenumber:   429070  
l_linenumber:        2  count_l_linenumber:  1285828  

<timing>
                          finish:    0.1 ms
                 totalKernelTime:    0.7 ms
                    krnl_reduce1:    0.0 ms
                    krnl_reduce2:    0.0 ms
</timing>
==13514== Profiling application: ./tpch23_advanced
==13514== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   54.15%  3.8083ms         4  952.08us  1.1200us  1.9043ms  [CUDA memcpy DtoH]
                   36.46%  2.5638ms         1  2.5638ms  2.5638ms  2.5638ms  [CUDA memcpy HtoD]
                    9.05%  636.74us         1  636.74us  636.74us  636.74us  krnl_lineitem1(int*, int*, int*, int*)
                    0.10%  7.1040us         3  2.3680us  2.2720us  2.4320us  void initArray<int>(int*, int, int)
                    0.09%  6.3680us         1  6.3680us  6.3680us  6.3680us  krnl_reduce1(int*, int*, agg_ht<apayl2>*, int*)
                    0.09%  6.3680us         1  6.3680us  6.3680us  6.3680us  krnl_reduce2(agg_ht<apayl2>*, int*, int*, int*, int*)
                    0.05%  3.6800us         1  3.6800us  3.6800us  3.6800us  void initAggHT<apayl2>(agg_ht<apayl2>*, int)
      API calls:   90.37%  189.44ms        14  13.531ms     989ns  188.67ms  cudaDeviceSynchronize
                    4.07%  8.5408ms         2  4.2704ms  4.2575ms  4.2833ms  cudaMallocHost
                    3.09%  6.4837ms         5  1.2967ms  11.932us  2.6335ms  cudaMemcpy
                    1.94%  4.0612ms         2  2.0306ms  2.0201ms  2.0412ms  cudaFreeHost
                    0.23%  481.93us         9  53.547us  2.9410us  102.04us  cudaMalloc
                    0.11%  234.94us         7  33.563us  2.7800us  80.375us  cudaFree
                    0.06%  119.08us        97  1.2270us      96ns  50.860us  cuDeviceGetAttribute
                    0.05%  109.65us         1  109.65us  109.65us  109.65us  cuDeviceTotalMem
                    0.04%  90.496us         7  12.928us  3.3690us  25.878us  cudaLaunchKernel
                    0.02%  31.576us         2  15.788us  12.471us  19.105us  cudaMemGetInfo
                    0.01%  18.220us         1  18.220us  18.220us  18.220us  cuDeviceGetName
                    0.00%  3.8240us         1  3.8240us  3.8240us  3.8240us  cuDeviceGetPCIBusId
                    0.00%  2.3480us        13     180ns      84ns     511ns  cudaGetLastError
                    0.00%     927ns         3     309ns     102ns     674ns  cuDeviceGetCount
                    0.00%     578ns         2     289ns     104ns     474ns  cuDeviceGet
                    0.00%     157ns         1     157ns     157ns     157ns  cuDeviceGetUuid
```

![image](https://user-images.githubusercontent.com/28505034/145019758-2d3509e8-c457-4446-872f-045d07043313.png)

**Comment by JigaoLuo** [comment]
If the perfect hash table size is known:

```
(base) jigao@jigao-System-Product-Name:~/Desktop/dogqc/bin$ ./23a
Memory 0.2 / 5.8 GB
Memory 0.2 / 5.8 GB
Shared memory usage: 3556 / 49152 bytes.

Result: 7 tuples
l_linenumber:        5  count_l_linenumber:   643287  
l_linenumber:        4  count_l_linenumber:   857015  
l_linenumber:        6  count_l_linenumber:   429070  
l_linenumber:        1  count_l_linenumber:  1500000  
l_linenumber:        7  count_l_linenumber:   214621  
l_linenumber:        3  count_l_linenumber:  1071394  
l_linenumber:        2  count_l_linenumber:  1285828  

<timing>
                          finish:    0.0 ms
                  krnl_lineitem1:    0.5 ms
                     krnl_reduce:    0.0 ms
                 totalKernelTime:    0.5 ms
</timing>
```
